### ISSUE TITLE
Implement GIVbacks claim window

### ISSUE THREAD

**Issue by laurenluz** [Giveth/GIVeconomy#912]
We might have some GIV from GIVbacks that has been distributed but left unclaimed for a long while. We should set up a process for reclaiming that GIV back to the DAO.

Is there a way to track how long GIV has been left unclaimed for to get an idea of what there is here?
Do we have a process for being able to reclaim that GIV or redirect it back to Giveth? Where do we send it back to? the rGIV DAO?

Creating this issue to start the conversation. I think we'll need a forum post for it.

@divine-comedian @aminlatifi @WhyldWanderer

**Comment by aminlatifi** [comment]
@laurenluz We can retrieve the list of givback recipients who have not claimed their tokens (have not claimed the liquidated part of their stream at all). It means, their stream is added up by GIVBack, but they have not interacted with TokenDistro to claim tokens directly or indirectly (like farming and harvesting reward)

There is a point in defining "How long GIV has been left unclaimed". Because potentially a user can receive givbacks in multiple rounds. If the user has not claimed any GIV, which round must be the origin time to calculate duration?

**Comment by laurenluz** [comment]
yeah good point @aminlatifi

is it easy/fast to get the info about how much GIV is allocated to addresses that have never interacted with the tokendistro to date? We could use it to decide next steps... if that number is in itself large, it is worth it to continue to define a clear system and communicate it to users.

As for "which round must be the origin time to calculate duration?" I Think we should consider the furthest back time as the "origin". so if a user interacted with the token distro since launch... but the last time was 1 year ago... we could consider relclaiming their GIV.

So we can basically define a system like "you need to claim/harvest/interact with the token distro at least one time per year, or your GIV will be sent back to the DAO".

**Comment by aminlatifi** [comment]
> yeah good point @aminlatifi
> 
> is it easy/fast to get the info about how much GIV is allocated to addresses that have never interacted with the tokendistro to date? We could use it to decide next steps... if that number is in itself large, it is worth it to continue to define a clear system and communicate it to users.
> 
> As for "which round must be the origin time to calculate duration?" I Think we should consider the furthest back time as the "origin". so if a user interacted with the token distro since launch... but the last time was 1 year ago... we could consider relclaiming their GIV.
> 
> So we can basically define a system like "you need to claim/harvest/interact with the token distro at least one time per year, or your GIV will be sent back to the DAO".

I will prepare a query to return users with allocated tokens who have never claimed their tokens. 
Regarding your example, it seems to explain "last time", not "furthest back time". Am I right?
Suppose a user has claimed once 2 years ago and once 1 year ago, the expected outcome must be 1 year. Right?

**Comment by aminlatifi** [comment]
@laurenluz I prepared this query to show users with allocated GIV who have not claimed any of it. 

https://dune.com/queries/3799716/6388453/


It also has separate columns for allocations associated with the explicit GIVBack pay calls. FYI, some allocations like this were not done with explicit givback pay method
https://gnosisscan.io/tx/0x586b61eec3677f1c2c9e54ca452ee7ea6deab0380a6747508648a8022c43e186#eventlog
@divine-comedian may know better why.

All above is done for gnosis network. If it's okay we can prepare the same for other chains.

**Comment by aminlatifi** [comment]
<img width="2616" alt="image" src="https://github.com/Giveth/GIVeconomy/assets/5684607/aa5c2f45-0c77-4dee-9dfc-4eee3e3003e8">


total_unclaimed | total_unclaimed_givback
-- | --
8382074.288546356 | 6961985.869643101

**Comment by divine-comedian** [comment]
> @laurenluz I prepared this query to show users with allocated GIV who have not claimed any of it.
> 
> https://dune.com/queries/3799716/6388453/
> 
> It also has separate columns for allocations associated with the explicit GIVBack pay calls. FYI, some allocations like this were not done with explicit givback pay method https://gnosisscan.io/tx/0x586b61eec3677f1c2c9e54ca452ee7ea6deab0380a6747508648a8022c43e186#eventlog @divine-comedian may know better why.
> 
> All above is done for gnosis network. If it's okay we can prepare the same for other chains.

The only users I could think of would be those who haven't claimed their praise rewards or vesting

**Comment by laurenluz** [comment]
wow this is great @aminlatifi!

8.4M GIV is actually quite a lot, and I think this might be worth the effort to make a system around. Let's chat about it in the GIVeconomy call today.

**Comment by laurenluz** [comment]
Hey @aminlatifi, as we discussed in the last GIVeconomy call:

We want to implement 2 conditions for if an address should have the GIV claimed back by the DAO:
1. Address should not have interacted with the token distro ever, AND
2. It has been at least 6 months (or 1 year) since the last time they were allocated any new GIV

We should include allocations from GIVbacks, praise & vesting.

Can you run some queries (and check the existing query for bugs) to compare also what is the GIV amount we could "claim back" if condition 2 considers the time frame to be 6 months vs. 1 year?

Once we have those GIV amounts, I can make a forum post & a vote so that we can implement this and send those tokens back to the GIVgarden multisig.

**Comment by aminlatifi** [comment]
> Hey @aminlatifi, as we discussed in the last GIVeconomy call:
> 
> We want to implement 2 conditions for if an address should have the GIV claimed back by the DAO:
> 
> 1. The address should not have interacted with the token distro ever, AND
> 2. It has been at least 6 months (or 1 year) since the last time they were allocated any new GIV
> 
> We should include allocations from GIVbacks, praise & vesting.
> 
> Can you run some queries (and check the existing query for bugs) to compare also what is the GIV amount we could "claim back" if condition 2 considers the time frame to be 6 months vs. 1 year?
> 
> Once we have those GIV amounts, I can make a forum post & a vote so that we can implement this and send those tokens back to the GIVgarden multisig.

@laurenluz I made [this query](https://dune.com/queries/3954471). The number of days since last allocated can be set with a parameter called `daysSinceLastAllocate`. You can change it and execute the query.
As I run it now, the total value we can get back to DAO is 
365 days: 4352354.1046977
180 days: 6575959.732925157

As an intro to the query, I queried all allocations to accounts that have not ever claimed from token distro. 
You said you are interested to allocations via givback, praise and vesting. There is no way for us to distinguish allocations for praise and vesting, only givback has explicit defined event on token distro. So, you can look at more details of allocations
[here](https://dune.com/queries/3799716)
<img width="2528" alt="image" src="https://github.com/user-attachments/assets/49c0d2ab-c11a-46d7-b4cf-1078766d5b93">

**Comment by laurenluz** [comment]
> As an intro to the query, I queried all allocations to accounts that have not ever claimed from token distro.
> You said you are interested to allocations via givback, praise and vesting. There is no way for us to distinguish allocations for praise and vesting, only givback has explicit defined event on token distro.

I think when we were discussing it, we said that we can't just decide to claim back "givbacks" because if we claim it back, we grab back their entire allocation. So we should consider basically all allocations from the token distro (not just ones market GIVbacks). Is that basically what you check @aminlatifi ? Or is the query only considering allocations from givbacks?

**Comment by aminlatifi** [comment]
> > As an intro to the query, I queried all allocations to accounts that have not ever claimed from token distro.
> > You said you are interested to allocations via givback, praise and vesting. There is no way for us to distinguish allocations for praise and vesting, only givback has explicit defined event on token distro.
> 
> I think when we were discussing it, we said that we can't just decide to claim back "givbacks" because if we claim it back, we grab back their entire allocation. So we should consider basically all allocations from the token distro (not just ones market GIVbacks). Is that basically what you check @aminlatifi ? Or is the query only considering allocations from givbacks?

It considers all. But I can only reports GIVbacks as well if you like

**Comment by laurenluz** [comment]
All is good! We can make a decision about 6 months of 1 year in the GIVeconomy call, and then move forward with this :-D

**Comment by laurenluz** [comment]
ok! forum post is up! so if this passes the gov process, we can proceed: https://forum.giveth.io/t/implementing-a-giv-claim-window-and-a-givbacks-minimum-donation-amount/1512

@aminlatifi we know how to "claim back" that GIV right? can you example how we could do it?

**Comment by aminlatifi** [comment]
> ok! forum post is up! so if this passes the gov process, we can proceed: https://forum.giveth.io/t/implementing-a-giv-claim-window-and-a-givbacks-minimum-donation-amount/1512
> 
> @aminlatifi we know how to "claim back" that GIV right? can you example how we could do it?

Yep, using this method on token distro
<img width="518" alt="image" src="https://github.com/user-attachments/assets/ede99fc7-80e1-46fd-b243-feb1b9153b94">

**Comment by laurenluz** [comment]
Ok this vote passed! https://snapshot.org/#/giv.eth/proposal/0x6bb75975839c993db2cddd9254fd18f3fb20624cca00551408e4e9fe3ee0a4f2

So we can implement the condition that:
- if the user's last allocation was 270 days ago
- and if the user had never claimed from the token distro

So we can execute this, and take all the streams and send them to the GIVgarden multisig instead... at least on Gnosis Chain.

@divine-comedian can you help to deploy the GIVgarden multisig on Optimism too, and then we can do on OP.

**Comment by laurenluz** [comment]
https://app.safe.global/settings/setup?safe=oeth:0xd10BAC02a02747cB293972f99981F4Faf78E1626

this is the givgarden safe on OP

**Comment by laurenluz** [comment]
and this is the givgarden safe on GNO: https://app.safe.global/settings/setup?safe=gno:0xd10BAC02a02747cB293972f99981F4Faf78E1626

I asked almond to prep a tweet warning people "last chance to claim your old GIV allocations", and once that goes out, I think we can wait another 5 days before actually redirecting this GIV to the garden multisigs. will update here once that timing is done.

cc @aminlatifi @divine-comedian

**Comment by laurenluz** [comment]
almond will post a tweet tomorrow that says that people have 1 more week to claim... so basically, we can execute on this on September 12, 2024 @aminlatifi

**Comment by laurenluz** [comment]
@aminlatifi this is ready to start up tomorrow

**Comment by laurenluz** [comment]
from the GIVeconomy call, there are about 3677 address we will need to "transfer allocation" for. Here are the results: https://dune.com/queries/3799716/6388453?daysSinceLastAllocate_t6c1ea=270

@aminlatifi can check how we can do these in bulk or batched

**Comment by laurenluz** [comment]
@mohammadranjbarz can you have a look?

**Comment by laurenluz** [comment]
@aminlatifi @mohammadranjbarz do we have time to revist this issue?

**Comment by mohammadranjbarz** [comment]
@aminlatifi Can you go forward for this issue ?
If you can't do it we can have a call, explain it to me, so I will take it

FYI @laurenluz

**Comment by aminlatifi** [comment]
> @aminlatifi Can you go forward for this issue ? If you can't do it we can have a call, explain it to me, so I will take it
> 
> FYI @laurenluz

@mohammadranjbarz Safe has ApiKit which allows creating multisig transaction(s) automatically. We would be able to create batch transactions as well, which will include multiple balance transfers in a single transaction.
There is more to read here.
https://docs.safe.global/sdk/api-kit/guides/propose-and-confirm-transactions

**Comment by mohammadranjbarz** [comment]
@aminlatifi This is two tx samples I create by the script, can you check it, if it's ok I can delete current TXs and execute the script again the create TXs


https://app.safe.global/transactions/queue?safe=gno:0xcF5FD2cC19C69C33310CEB740F10BB10091211f3



https://github.com/mohammadranjbarz/gnosis-safe-batch-tx

Also you can run this script to decode the data to see if tx payload data is correct or not

https://github.com/mohammadranjbarz/gnosis-safe-batch-tx/blob/master/decode-data.ts

**Comment by mohammadranjbarz** [comment]
@laurenluz 
I created the TXs on gnosis safe

https://app.safe.global/transactions/queue?safe=gno:0xcF5FD2cC19C69C33310CEB740F10BB10091211f3

**Comment by mohammadranjbarz** [comment]
@laurenluz  @aminlatifi 
Can you check it before signing it, for instance, is the number of addresses is correct (`daysSinceLastAllocate: 270`) , **4700**

**Comment by laurenluz** [comment]
@mohammadranjbarz I think there is some issue here... it looks like you set every batch with the same nonce? so executing one will replace the others

![Image](https://github.com/user-attachments/assets/2d5f3490-cc35-4943-a7e9-493a8a4546c1)

how do I check the correct number of addresses? @aminlatifi ?

btw guys if we have something like this in a multisig that REALLY needs review... we should make a message to signers and ask them to holdoff explicity until it is confirmed by someone specific... otherwise they may be accidentally pushed through.

**Comment by laurenluz** [comment]
Also noticing that this is the Giveth "reward board" multisig. Is that the correct one? I didn't realize the reward board multisig was an admin controlling GIV allocations from the token distro

**Comment by laurenluz** [comment]
@mohammadranjbarz giv.eth is the multisig where the txs can be created. Can you create this tx with just the 1st 10 addresses so we can verify that it works properly in the correct sig?

**Comment by laurenluz** [comment]
@mohammadranjbarz just a poke on this now that you're done w/ some of unicorn. can you pick it up again?

**Comment by mohammadranjbarz** [comment]
> [@mohammadranjbarz](https://github.com/mohammadranjbarz) just a poke on this now that you're done w/ some of unicorn. can you pick it up again?

Yes of course will pick it, just in the coming week, because after the unicorn release, I was a little busy with the AI Agenting

**Comment by divine-comedian** [comment]
@laurenluz is this still on your radar? what we should we do with this one?

**Comment by laurenluz** [comment]
@aminlatifi pinging you here to investigate

**Comment by aminlatifi** [comment]
> [@aminlatifi](https://github.com/aminlatifi) pinging you here to investigate

I got the code @mohammadranjbarz had developed with configurations. 
@laurenluz Please put here any configuration you need to apply, e.g. multisig address, ...

**Comment by laurenluz** [comment]
Not sure I understand your request @aminlatifi 

From my understanding, shouldn't only the admin of the token distro be able to "claw back" GIV allocations? So it would by default just be giv.eth...

Maybe you can explain to me exactly what you're planning to do. At this point, with so many miscommunications and time elapsed on this issue, I'm not sure we're on the same page anymore.

**Comment by aminlatifi** [comment]
> Not sure I understand your request [@aminlatifi](https://github.com/aminlatifi)
> 
> From my understanding, shouldn't only the admin of the token distro be able to "claw back" GIV allocations? So it would by default just be giv.eth...
> 
> Maybe you can explain to me exactly what you're planning to do. At this point, with so many miscommunications and time elapsed on this issue, I'm not sure we're on the same page anymore.

Yeah, the admin can do it. Who must be the recipient? givbacks must be transferred to which address?

**Comment by laurenluz** [comment]
I think we should send them back to the givgarden multisigs. that's not an issue right?I think givgarden on gnosis already has a stream. As for givgarden on optimism, I'm not sure.

the address is the same on both chains

[Gnosis: 0xd10BAC02a02747cB293972f99981F4Faf78E1626](https://gnosisscan.io/address/0xd10BAC02a02747cB293972f99981F4Faf78E1626)
[Optimsim: 0xd10BAC02a02747cB293972f99981F4Faf78E1626](https://optimistic.etherscan.io/address/0xd10BAC02a02747cB293972f99981F4Faf78E1626)

**Comment by laurenluz** [comment]
@aminlatifi just following up here

**Comment by aminlatifi** [comment]
> [@aminlatifi](https://github.com/aminlatifi) just following up here

@laurenluz I proposed transactions using delegate wallet, so no one is signed. You can see them here.
https://app.safe.global/transactions/queue?safe=gno:0x4D9339dd97db55e3B9bCBE65dE39fF9c04d1C2cd

The transactions are supposed to transfer allocations to "0xd10BAC02a02747cB293972f99981F4Faf78E1626"
